File: backend/internal/query/query.go
// Copyright (C) 2025 Austin Beattie
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

package query

import (
	"errors"

	"github.com/octobud-hq/octobud/backend/internal/db"
	"github.com/octobud-hq/octobud/backend/internal/query/parse"
	"github.com/octobud-hq/octobud/backend/internal/query/sql"
)

// Error definitions
var (
	ErrTokenizationFailed  = errors.New("tokenization failed")
	ErrParseFailed         = errors.New("parse failed")
	ErrSQLGenerationFailed = errors.New("SQL generation failed")
)

// Node is an alias for parse.Node for convenience
type Node = parse.Node

// ParseAndValidate parses a query string and validates it
// Returns the AST node if successful
func ParseAndValidate(queryStr string) (Node, error) {
	if queryStr == "" {
		return nil, nil
	}

	lexer := parse.NewLexer(queryStr)
	tokens, err := lexer.Tokenize()
	if err != nil {
		return nil, errors.Join(ErrTokenizationFailed, err)
	}

	parser := parse.NewParser(tokens)
	ast, err := parser.Parse()
	if err != nil {
		return nil, errors.Join(ErrParseFailed, err)
	}

	validator := parse.NewValidator()
	if err := validator.Validate(ast); err != nil {
		return nil, errors.Join(parse.ErrValidationFailed, err)
	}

	return ast, nil
}

// BuildQuery parses a query string and generates SQL with explicit query context:
// - Empty query "" → Default inbox: exclude archived, snoozed (active), muted, filtered (backward compatibility)
// - Query with in: operator (any value) → No defaults (in: operator explicitly handles lifecycle)
// - Query without in: operator (non-empty) → Apply muted-only default (exclude muted unless explicitly requested)
func BuildQuery(queryStr string, limit, offset int32) (db.NotificationQuery, error) {
	return BuildQueryWithOptions(queryStr, limit, offset, true)
}

// BuildQueryWithOptions parses a query string and generates SQL with unified business rules
func BuildQueryWithOptions(
	queryStr string,
	limit, offset int32,
	includeSubject bool,
) (db.NotificationQuery, error) {
	ast, err := ParseAndValidate(queryStr)
	if err != nil {
		return db.NotificationQuery{}, err
	}

	builder := sql.NewBuilder()
	query, err := builder.Build(ast)
	if err != nil {
		return db.NotificationQuery{}, errors.Join(ErrSQLGenerationFailed, err)
	}

	// Apply unified business rules
	query = applyUnifiedDefaults(query, ast, queryStr)

	query.Limit = limit
	query.Offset = offset
	query.IncludeSubject = includeSubject

	return query, nil
}

// applyUnifiedDefaults applies default filters based on explicit query context:
// - Empty query → Default inbox: exclude archived, snoozed (active), muted, filtered (backward compatibility)
// - Query with in: operator (any value) → No defaults (in: operator explicitly handles lifecycle)
// - Query without in: operator (non-empty) → Apply muted-only default (exclude muted unless explicitly requested)
func applyUnifiedDefaults(
	query db.NotificationQuery,
	ast Node,
	queryStr string,
) db.NotificationQuery {
	// 1. Empty query → Default inbox: exclude archived, snoozed (active), muted, filtered (backward compatibility)
	if queryStr == "" || ast == nil {
		return ApplyInboxDefaults(query)
	}

	// 2. Query with in: operator (any value) → No defaults (in: operator explicitly handles lifecycle)
	// This includes in:inbox, in:archive, in:snoozed, in:filtered, in:anywhere, etc.
	if parse.HasInOperator(ast) {
		return query
	}

	// 3. Query without in: operator (non-empty) → Apply muted-only default
	// This allows archived, snoozed, and filtered to show unless explicitly excluded
	// BUT: if query explicitly asks for muted (is:muted or muted:true), don't apply the default
	if parse.HasExplicitMuted(ast) {
		return query
	}
	return ApplyMutedOnlyDefaults(query)
}

File: extension/dev/stub-server/query.mjs
// Copyright (C) 2025 Austin Beattie
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

/**
 * A deliberately small stand-in for backend/internal/query.
 *
 * The real engine is a lexer -> parser -> AST -> SQL builder supporting
 * parentheses, explicit AND/OR/NOT and free-text search across several columns.
 * This handles only what the panel's views actually need: space-separated terms
 * ANDed together, comma-separated values ORed within one field, and a leading
 * `-` for negation.
 *
 * It exists so the seeded views return visibly different result sets. It is not
 * a reimplementation, and it should not grow into one — if a test needs richer
 * query semantics, run the real backend instead.
 */

const BOOLEAN_FIELDS = new Set(["starred", "archived", "snoozed", "muted", "filtered", "read"]);

function normalizeType(value) {
	const lowered = String(value ?? "").toLowerCase();
	if (lowered === "pullrequest" || lowered === "pull_request") return "pull_request";
	return lowered;
}

function contains(haystack, needle) {
	return String(haystack ?? "")
		.toLowerCase()
		.includes(String(needle ?? "").toLowerCase());
}

function isInInbox(item) {
	return !item.archived && !item.snoozedUntil && !item.filtered;
}

/** Evaluates a single `field:value` pair (already lowercased field). */
function matchesPair(item, field, value) {
	const lowered = String(value).toLowerCase();

	switch (field) {
		case "is":
			switch (lowered) {
				case "read":
					return item.isRead;
				case "unread":
					return !item.isRead;
				case "starred":
					return item.starred;
				case "snoozed":
					return Boolean(item.snoozedUntil);
				case "archived":
					return item.archived;
				case "muted":
					return item.muted;
				case "filtered":
					return item.filtered;
				default:
					return false;
			}

		case "in":
			switch (lowered) {
				case "inbox":
					return isInInbox(item);
				case "archive":
					return item.archived;
				case "snoozed":
					return Boolean(item.snoozedUntil);
				case "filtered":
					return item.filtered;
				case "anywhere":
					return true;
				default:
					return false;
			}

		case "type":
			return normalizeType(item.subjectType) === normalizeType(lowered);

		case "repo":
			return contains(item.repository?.fullName, lowered);

		case "reason":
			return contains(item.reason, lowered);

		case "author":
			return contains(item.authorLogin, lowered);

		case "tags":
			return (item.tags ?? []).some(
				(tag) => tag.slug.toLowerCase() === lowered || contains(tag.name, lowered)
			);

		// `starred:true` style. The frontend's built-in view constants use this
		// form while the backend's system views use `is:`, so both are accepted.
		case "starred":
		case "archived":
		case "muted":
		case "filtered":
		case "read": {
			const actual =
				field === "read" ? item.isRead : field === "starred" ? item.starred : item[field];
			return lowered === "true" ? Boolean(actual) : !actual;
		}
		case "snoozed":
			return lowered === "true" ? Boolean(item.snoozedUntil) : !item.snoozedUntil;

		default:
			return false;
	}
}

/** Free text matches title, repo or author, mirroring the real engine's spirit. */
function matchesFreeText(item, text) {
	return (
		contains(item.subjectTitle, text) ||
		contains(item.repository?.fullName, text) ||
		contains(item.authorLogin, text)
	);
}

export function parseQuery(queryString) {
	return String(queryString ?? "")
		.split(/\s+/)
		.map((raw) => raw.trim())
		.filter(Boolean)
		.map((raw) => {
			const negated = raw.startsWith("-");
			const body = negated ? raw.slice(1) : raw;
			const separator = body.indexOf(":");

			if (separator === -1) {
				return { negated, field: null, values: [body] };
			}

			return {
				negated,
				field: body.slice(0, separator).toLowerCase(),
				values: body
					.slice(separator + 1)
					.split(",")
					.filter(Boolean),
			};
		});
}

function matchesTerm(item, term) {
	const matched = term.field
		? term.values.some((value) => matchesPair(item, term.field, value))
		: term.values.some((value) => matchesFreeText(item, value));
	return term.negated ? !matched : matched;
}

/**
 * Mirrors the implicit defaults in backend/internal/query/query.go: an empty
 * query means the inbox, a query that scopes itself with `in:` gets nothing
 * added, and anything else just hides muted items.
 */
function applyImplicitDefaults(item, terms) {
	if (terms.length === 0) {
		return isInInbox(item) && !item.muted;
	}
	if (terms.some((term) => term.field === "in")) {
		return true;
	}
	return !item.muted;
}

export function evaluateQuery(items, queryString) {
	const terms = parseQuery(queryString);
	return items.filter(
		(item) => applyImplicitDefaults(item, terms) && terms.every((term) => matchesTerm(item, term))
	);
}
